test/stress/stress_test_upstream.h

```diff
@@ -198,50 +198,65 @@ class LocalListenSocket : public Network::TcpListenSocket {
+  COUNTER(accepts)                                                                                 \
+  COUNTER(requests_received)                                                                       \
+  COUNTER(local_closes)                                                                            \
+  COUNTER(remote_closes)
+
 /**
- * A convenience class for passing callbacks to a Server. If no callbacks are
- * provided, default callbacks that track some simple metrics will be used. If
- * callbacks are provided, they will be wrapped with callbacks that maintain the
- * same simple set of metrics.
+ * A convenience class for passing callbacks to a Server. If no callbacks are provided, default
+ * callbacks that track some simple metrics will be used. If callbacks are provided, they will be
+ * wrapped with callbacks that maintain the same simple set of metrics.
  */
 class ServerCallbackHelper {
 public:
+  struct Stats {
+    ALL_SERVER_HELPER_STATS(GENERATE_COUNTER_STRUCT)
+  };
+
   explicit ServerCallbackHelper(ServerRequestCallback&& request_callback = nullptr,
                                 ServerAcceptCallback&& accept_callback = nullptr,
                                 ServerCloseCallback&& close_callback = nullptr);
   ServerCallbackHelper(const ServerCallbackHelper&) = delete;
   ServerCallbackHelper& operator=(const ServerCallbackHelper&) = delete;
   virtual ~ServerCallbackHelper() = default;
 
-  uint32_t connectionsAccepted() const;
-  uint32_t requestsReceived() const;
-  uint32_t localCloses() const;
-  uint32_t remoteCloses() const;
   ServerAcceptCallback acceptCallback() const;
   ServerRequestCallback requestCallback() const;
   ServerCloseCallback closeCallback() const;
 
   /*
-   * Wait until the server has accepted n connections and seen them closed (due
-   * to error or client close)
+   * Wait until the server has accepted n connections and seen them closed (due to error or client
+   * close)
+   *
+   * @param connections_to_close Wait for this number of connections to close before returning.
    */
-  void wait(uint32_t connections);
+  void wait(uint32_t connections_to_close);
 
   /*
    * Wait until the server has seen a close for every connection it has
    * accepted.
    */
   void wait();
 
+  /**
+   * Access the metrics maintained by the server.
+   *
+   * @return The server's metrics
+   */
+  const Stats& stats() const;
+
+  void initStats(std::unique_ptr<Stats>&& stats);
+
 private:
   ServerAcceptCallback accept_callback_;
   ServerRequestCallback request_callback_;
   ServerCloseCallback close_callback_;
-
-  std::atomic<uint32_t> accepts_{0};
-  std::atomic<uint32_t> requests_received_{0};
-  uint32_t local_closes_{0};
-  uint32_t remote_closes_{0};
+  // Counters needed for control flow are not implemented as stats.
+  uint32_t connections_accepted_{0};
+  uint32_t connections_closed_{0};
+  std::unique_ptr<Stats> stats_;
```

Comment:
I would put the mutex_ above the fields that it is protecting and add a GUARDED_BY annotation for those fields.

Reply:
Add mutex_.AssertHeld() in the callback.